robot-server/robot_server/runs/run_view.py

```diff
@@ -0,0 +1,129 @@
+"""Run response model factory."""
+from dataclasses import replace
+from datetime import datetime
+from typing import Any, Dict, List, Optional, Tuple
+
+from opentrons.protocol_engine import (
+    Command as ProtocolEngineCommand,
+    EngineStatus,
+    LoadedLabware,
+    LoadedPipette,
+)
+
+from .run_store import RunResource
+from .action_models import RunAction, RunActionCreateData
+from .run_models import (
+    Run,
+    RunCreateData,
+    BasicRun,
+    BasicRunCreateData,
+    ProtocolRun,
+    ProtocolRunCreateData,
+    RunCommandSummary,
+)
+
+
+class RunView:
+    """Interface to build run model instances from data.
+
+    Resources consumed and returned by this class will be treated as
+    immutable.
+    """
+
+    @staticmethod
+    def as_resource(
+        run_id: str,
+        created_at: datetime,
+        create_data: Optional[RunCreateData],
+    ) -> RunResource:
+        """Create a new run resource instance from its create data.
+
+        Arguments:
+            run_id: Unique identifier.
+            created_at: Resource creation timestamp.
+            create_data: Data used to create the run.
+
+        Returns:
+            The run in its internal resource representation, for use in
+                the `RunStore` and other classes.
+        """
+        return RunResource(
+            run_id=run_id,
+            created_at=created_at,
+            create_data=create_data or BasicRunCreateData(),
+            actions=[],
+        )
+
+    @staticmethod
+    def with_action(
+        run: RunResource,
+        action_id: str,
+        action_data: RunActionCreateData,
+        created_at: datetime,
+    ) -> Tuple[RunAction, RunResource]:
+        """Create a new run control action resource instance.
+
+        Arguments:
+            run: The run resource to add the command to.
+            action_id: Unique ID to assign to the command resource.
+            action_data: Data used to create the command resource.
+            created_at: Resource creation timestamp.
```

Comment:
I think these should be talking about actions?

```suggestion
        Arguments:
            run: The run resource to add the action to.
            action_id: Unique ID to assign to the action resource.
            action_data: Data used to create the action resource.
            created_at: Resource creation timestamp.
```